### PR TITLE
[WIP] initial commit nvFuser intro

### PR DESCRIPTION
**Pull request by ptrblck**
placeholder PR for review

### CODE REVIEW

**Comment by ptrblck** [comment]
CC @csarofeen @kevinstephano

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
+# -*- coding: utf-8 -*-
+"""
+Getting started with nvFuser
+****************************
+
+**Authors**: `Christian Sarofeen <https://github.com/csarofeen>`_
+`Kevin Stephano <https://github.com/kevinstephano>`_
+`Piotr Bialecki <https://github.com/ptrblck>`_
+`Neal Vaidya`
```
Comment (ptrblck) [inline intermediate_source/nvfuser_intro_tutorial.py]: Add more authors?

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+
+
+Introduction
+------------
+
+This tutorial shows how to trace PyTorch programs using TorchScript, FuncTorch,
+and TorchDynamo, and then execute them with nvFuser.
```
Comment (ptrblck) [inline intermediate_source/nvfuser_intro_tutorial.py]: remove `TorchDynamo`

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+Getting started with nvFuser
+****************************
+
+**Authors**: `Christian Sarofeen <https://github.com/csarofeen>`_
+`Kevin Stephano <https://github.com/kevinstephano>`_
+`Piotr Bialecki <https://github.com/ptrblck>`_
+`Neal Vaidya`
+
+
+Introduction
```
Comment (ptrblck) [inline intermediate_source/nvfuser_intro_tutorial.py:14]: Add more text to introduction section

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+------------
+
+This tutorial shows how to trace PyTorch programs using TorchScript, FuncTorch,
+and TorchDynamo, and then execute them with nvFuser.
+
+Importing Packages and Selecting a Device
+-----------------------------------------
+In order to run this tutorial and see the benefits of using nvfuser, 
+you would need to install the `1.12.0` PyTorch release as well as
```
Comment (ptrblck) [inline intermediate_source/nvfuser_intro_tutorial.py]: should we add a check here?

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+import random
+random.seed(42)
+import time
+from typing import List
+from copy import deepcopy
+
+import torch
+import torch.nn.functional as F
+
+from torch.utils import collect_env
```
Comment (ptrblck) [inline intermediate_source/nvfuser_intro_tutorial.py]: might not be needed, if we check the version

File: index.rst [hunk 6]
```diff
@@ -490,6 +490,13 @@ What's new in PyTorch tutorials?
    :link: intermediate/torchserve_with_ipex
    :tags: Model-Optimization,Production
 
+.. customcarditem::
+   :header: Introduction to nvFuser
+   :card_description: An introduction to nvFuser TODO
+   :image: _static/img/thumbnails/cropped/60-min-blitz.png
+   :link: intermediate/nvfuser_intro_tutorial
+   :tags: Model-Optimization
```
Comment (ptrblck) [inline index.rst:498]: Added this intro tutorial to `Model-Optimization` as it seems to be the best fit

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+# Transformer Block With a Novel Normalization
+# ----------------------------------------------------
+# The ability to quickly execute chains of simple operations is
+# important as not every operation has a composite operation defined
+# in PyTorch. Previously, this meant researchers either had to define
+# an entirely new operation in PyTorch – which takes a lot of time and
+# knowledge of the lower-level PyTorch code as well as parallel
+# programming – or writing the operation in simpler PyTorch ops and
+# settling for poor performance. For example, let's replace LayerNorm
+# in our example with RMSNorm. Even though RMSNorm is a bit simpler
```
Comment (crcrpar) [inline intermediate_source/nvfuser_intro_tutorial.py:575]: qq: should we explain RMSNorm itself or add some reference for RMSNorm or is it known enough now?

File: index.rst [hunk 8]
```diff
@@ -490,6 +490,13 @@ What's new in PyTorch tutorials?
    :link: intermediate/torchserve_with_ipex
    :tags: Model-Optimization,Production
 
+.. customcarditem::
+   :header: Introduction to nvFuser
+   :card_description: An introduction to nvFuser TODO
```
Comment (crcrpar) [inline index.rst]:
```suggestion
   :card_description: An introduction to nvFuser
```

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+
+######################################################################
+# The Transformer Block
+# ---------------------
+# The network topology we’re going to focus on is the Transformer
+# Block for networks like BERT. As of writing this blog post nvFuser
+# provides acceleration of pointwise, reduction, and normalization
+# operations. These simple operations are the backbone of large
+# networks, so improving the speed of these operations can improve
+# overall network training speed. Future releases nvFuser will
```
Comment (crcrpar) [inline intermediate_source/nvfuser_intro_tutorial.py]: Does "Future releases nvFuser will" need some preposition (a) before "Future" or (b) between "releases" and "nvFuser"?

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,647 @@
...
+# It’s important to be very careful when measuring performance on the
+# GPU, as we want to remove any initialization costs and need
+# synchronization since it’s an asynchronous device. Since we will
+# measure many variations of this problem with and without nvFuser we
+# define a helper method called `profile_workload` and will use
+# `functool.partial` to concisely profile the workload.
+#
+
+# Utility to profile the workload
+def profile_workload(forward_func, grad_output, iteration_count=100):
```
Comment (nealvaidya) [inline intermediate_source/nvfuser_intro_tutorial.py]: Can we add a "label" parameter that gets printed with the output (like "eager" or "torchscript")? That will make it easier to compare the outputs of workloads with different settings.

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,647 @@
...
+# No changes from before are required for running with TorchScript, we
+# simply reuse the previous definition that we wrapped in
+# `torch.jit.script`.
+#
+# We’ll start as usual by performing some warm-up iterations, however
+# we won’t show nvFuser all of the input sizes, we’ll only show one
+# size for the warm-up.
+#
+
+# Perform warm-up iterations
```
Comment (nealvaidya) [inline intermediate_source/nvfuser_intro_tutorial.py:316]: Do we even need to perform the warm-up iterations here? The scripted definition is already compiled from the static example before, it doesn't need recompilation.

File: intermediate_source/nvfuser_intro_tutorial.py (hunk 7 above)
Comment (ptrblck) [reply intermediate_source/nvfuser_intro_tutorial.py:575]: Sounds like a good idea. Would this be the reference? https://arxiv.org/abs/1910.07467?context=cs

File: intermediate_source/nvfuser_intro_tutorial.py (hunk 9 above)
Comment (ptrblck) [reply intermediate_source/nvfuser_intro_tutorial.py]: I go with b):
> Future releases *of* nvFuser...

File: intermediate_source/nvfuser_intro_tutorial.py (hunk 11 above)
Comment (ptrblck) [reply intermediate_source/nvfuser_intro_tutorial.py:316]: Yes, I think so as your device might go back to P8 (sleep/idle) which would add a slight overhead in the fist CUDA op to push it back to P0 (perf. mode).

File: intermediate_source/nvfuser_intro_tutorial.py (hunk 7 above)
Comment (crcrpar) [reply intermediate_source/nvfuser_intro_tutorial.py:575]: I think so according to https://github.com/NVIDIA/apex/issues/1271, though, https://doi.org/10.48550/arXiv.1910.07467 might be better

File: intermediate_source/nvfuser_intro_tutorial.py (hunk 10 above)
Comment (ptrblck) [reply intermediate_source/nvfuser_intro_tutorial.py]: Done!

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,679 @@
...
+    dropout_prob=0.1,
+)
+profile_workload(
+    func, grad_output, iteration_count=100, label="TorchScript - Composite definition"
+)
+
+######################################################################
+# Before we get to the results, it is important to mention here that
+# nvFuser does not generate the exact same sequence of random numbers,
+# as random number generation in PyTorch is dependent on the precice
```
Comment (IvanYashchuk) [inline intermediate_source/nvfuser_intro_tutorial.py]:
```suggestion
# as random number generation in PyTorch is dependent on the precise
```

File: intermediate_source/nvfuser_intro_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,679 @@
...
+designed to meet the unique requirements of the PyTorch community,
+and it supports diverse network architectures and programs with
+dynamic inputs of varying shapes and strides. If you’re unfamiliar
+with nvFuser please review [this blog post, INSERT BLOG POST LINK].
+
+Importing Packages and Selecting a Device
+-----------------------------------------
+In order to run this tutorial and see the benefits of using nvfuser,
+you would need to install the `1.12.0` PyTorch release as well as
+`functorch` `0.2` or newer version of them.
```
Comment (IvanYashchuk) [inline intermediate_source/nvfuser_intro_tutorial.py]: I installed functorch using `pip install --user "git+https://github.com/pytorch/functorch.git"` and it seems it doesn't include `networkx` needed for `memory_efficient_fusion` 
```
RuntimeError: Need networkx installed to perform smart recomputation heuristics
```
I think the dependency on `networkx` should be included here.

**Pull request closed without merging**